KAFKA-7772: Dynamically Adjust Log Levels in Connect

wicknicks opened the pull request:
This PR implements changes proposed in [KIP-495](https://cwiki.apache.org/confluence/display/KAFKA/KIP-495%3A+Dynamically+Adjust+Log+Levels+in+Connect).

This change will introduce an `/admin/loggers` endpoint to the Connect worker that can be used to get/modify the log levels of any named logger in the Connect worker.

The following new configs are introduced: 

* `admin.listeners` to control where the admin endpoint will be made available. The default value should bring the `/admin` endpoint along with the existing endpoints. Setting it to empty disables the /admin endpoint altogether. Any other host:port string create that listener and only bring the admin endpoints on it.
* The `admin.listeners.https.` prefix is used to find SSL props for any HTTPS endpoints. 

Multiple tests are added to verify adding admin endpoints with various listeners, and to check if the REST endpoints behave as intended. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

wicknicks committed:
KAFKA-7772: Initial commit

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Start admin context only if admin endpoints are on different listeners

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Implementation of the logging resource

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Add /admin prefix to resource

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Add integration tests for LoggerResource

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Rename and clean up adminUrl method in RestServer

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Make checkstyle happy

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Align adminUrl to the contract with admin.listeners

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Make checkstyle gods happy

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Clean up code and add some javadoc

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -186,6 +186,12 @@
+    protected static final String ADMIN_LISTENERS_DOC = "List of comma-separated URIs the Admin REST API will listen on." +
+            " The supported protocols are HTTP and HTTPS." +
+            " An empty or blank string will disable this feature." +
+            " The default behavior is to use the regular listener (specified by the 'listeners' property).";
```
It'd be good to include the default here, even if it is empty. (See line 187, for example.)

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -114,24 +119,39 @@ public RestServer(WorkerConfig config) {
      * Adds Jetty connector for each configured listener
      */
-    public void createConnectors(List<String> listeners) {
+    public void createConnectors(List<String> listeners, List<String> adminListeners) {
```
Do we want to also keep a method with the old signature for backward compatibility?

wicknicks replied:
this method is only used in the line above, and is not part of the public API. are you wondering if any connectors are using this API since it is available on the classpath?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -114,24 +119,39 @@ public RestServer(WorkerConfig config) {
         jettyServer.setConnectors(connectors.toArray(new Connector[connectors.size()]));
+
+        if (adminListeners != null && !adminListeners.isEmpty()) {
+            for (String adminListener : adminListeners) {
```
Should we handle the case when `adminListener` is empty or blank? See line 126 above.

wicknicks replied:
added a validator that should exclude empty or blank listeners.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -150,10 +170,14 @@ public Connector createConnector(String listener) {
+        }
+
+        if (name != null) {
+            connector.setName(name);
```
Wouldn't it be much easier to do the following:
```
    public Connector createConnector(String listener, String name) {
        ...
        String hostname = ...
        int port = ...
        ...
        if (name == null || name.trim().isEmpty()) {
            name = String.format("%s_%s%d", PROTOCOL_HTTPS, hostname, port);
        }
        connector.setName(name);
        ...
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -298,6 +304,8 @@ protected static ConfigDef baseConfigDef() {
                 .define(REST_EXTENSION_CLASSES_CONFIG, Type.LIST, "",
                         Importance.LOW, REST_EXTENSION_CLASSES_DOC)
+                .define(ADMIN_LISTENERS_CONFIG, Type.LIST, null,
+                        Importance.LOW, ADMIN_LISTENERS_DOC)
```
Do we need a validator here that ensures that the values are one of the following?
1) null
2) an empty list
3) a list of non-empty strings

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -198,13 +223,39 @@ public void initializeResources(Herder herder) {
+            log.info("Adding admin resources to admin listener");
+            adminResourceConfig = new ResourceConfig();
+            adminResourceConfig.register(new JacksonJsonProvider());
+            addAdminResources(adminResourceConfig);
```
Why have an `addAdminResources(...)` method when it could simply call `register(...)` like the previous and next lines? The method is private, so it's not useful for subclasses.

I understand that we'd be duplicating some code, but it seems more complex to have the method abstract the one call.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -198,13 +223,39 @@ public void initializeResources(Herder herder) {
+        ServletHolder adminServletHolder = new ServletHolder(new ServletContainer(adminResourceConfig));
+        adminContext.setContextPath("/");
+        adminContext.addServlet(adminServletHolder, "/*");
+        adminContext.setVirtualHosts(new String[]{"@Admin"});
```
It is okay to do this if we're adding admin listeners to the main listener?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -223,16 +274,28 @@ public void initializeResources(Herder herder) {
 
+        if (adminResourceConfig != resourceConfig) {
+            try {
+                adminContext.start();
```
Might be good to add a debug log message right before this.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java:
```diff
@@ -287,6 +350,36 @@ else if (serverConnector != null && serverConnector.getPort() > 0)
+        UriBuilder builder = UriBuilder.fromUri(jettyServer.getURI());
+        builder.port(adminConnector.getLocalPort());
+
+        log.info("Admin URI: {}", builder.build());
```
Do we really want to log this every time this getter method is called?

wicknicks replied:
this method is only called once in `RestServer#initializeServer` and in integration tests to get the admin endpoints (so once per process). it is similar to how we log this info in `advertisedUrl()`.

rhauch replied:
Yes, the tests call it a lot, and we're adding an `Connect.adminUrl()` method that calls this, so it's possible other calls can be added in the future.
I'm not sure why this is even needed, given the existing log message on [line 219](https://github.com/apache/kafka/pull/7403/files#diff-2bc823e3987e8791b5c4a6818d455c66R219) -- which is where I would have expected it.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,173 @@
+            if (current.getLevel() != null) {
+                loggers.put(current.getName(), levelToMap(current.getLevel()));
+            }
+        }
```
Would it be a bit more obvious what's happening if we use Java streams?
```
Collections.list(enumeration)
           .stream()
           .filter(logger -> logger.getLevel() != nul)
           .forEach(logger -> loggers.put(logger.getName(), levelToMap(logger.getLevel());
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,173 @@
+                level = String.valueOf(logger.getLevel());
+            }
+
+            levelMap.put("level", level);
```
Could this be replaced with the following?
```
Map<String, String> levelMap = levelToMap(logger.getLevel());
```

Or, better yet, if `levelToMap(...)` were changed to take a `Level` and encapsulated the logic of whether the level was set, then we could clean up other places where this method is used and this code:
```
Map<String, String> levelMap = levelToMap(logger);
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,173 @@
+    public Response setLevel(final @PathParam("logger") String namedLogger,
+                             final Map<String, String> levelMap) {
+        String desiredLevelStr = levelMap.get("level");
+        Objects.requireNonNull(desiredLevelStr, "desired level was not specified in PUT request.");
```
Do we want to fail with a NPE if the payload does not contain `level`? Would it be better to use `BadRequestException`? Also, the error message should be a bit more clear about exactly what was missing.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,173 @@
+
+        List<Logger> childLoggers;
+        if (ROOT_LOGGER_NAME.equals(namedLogger)) {
+            childLoggers = Collections.list((Enumeration<Logger>) LogManager.getCurrentLoggers());
```
Seems like it would be good to have a unit test and is less expensive to test, and in order to do that this class could abstract out the calls to LogManager so that it could be stubbed or mocked in unit tests. It might also be useful to hide the use of Enumeration, especially if that means this class' methods could use more Java 8 streams.

rhauch reviewed (commented):
Nice progress, @wicknicks! A few comments below.

You asked about whether it's beneficial to add unit tests since we really need/want integration tests to really verify the functionality of the REST service implementation. I actually think it would be beneficial to *also* have a unit test for the `LoggingResource` class because the unit tests can be more thorough and lightweight (e.g., we can have many more tests that assert specific conditions, including checking for null or verifying the behavior when the inputs are not correct).

wicknicks committed:
KAFKA-7772: Address reviewer comments

Moved integration tests to RestServerTest.
Added SSL configs to the admin HTTPS listener.

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Add unit tests for LoggingResource

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-7772: Clean up RestServerTest

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@rhauch addressed most of your comments. couple of questions on others. this is ready for another pass. thanks!

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java:
```diff
@@ -375,4 +386,31 @@ public WorkerConfig(ConfigDef definition, Map<String, String> props) {
+                if (!(item instanceof String)) {
+                    throw new ConfigException("Invalid type for admin listener (expected String).");
+                }
+                if (((String) item).length() == 0) {
```
What about the following to also prevent whitespace-only strings:
```suggestion
                if (((String) item).trim().isEmpty()) {
```

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,200 @@
+     * @param logger a non-null log4j logger
+     * @return a singleton map whose key is level and the value is the string representation of the logger's effective log level.
+     */
+    private static Map<String, String> effLevelToMap(Logger logger) {
```
It's better to avoid abbreviations. Assuming "eff" means "effective", then maybe:
```suggestion
    private static Map<String, String> effectiveLevelToMap(Logger logger) {
```

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConfigTest.java:
```diff
@@ -0,0 +1,67 @@
+    @Test(expected = ConfigException.class)
+    public void testAdminListenersNotAllowingEmptyStrings() {
+        Map<String, String> props = baseProps();
+        props.put(WorkerConfig.ADMIN_LISTENERS_CONFIG, "http://a.b:9999,");
```
Maybe add a similar unit test to check using blank strings.

wicknicks replied:
`testAdminListenersNotAllowingBlankStrings` added below

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResourceTest.java:
```diff
@@ -0,0 +1,176 @@
+        when(loggingResource.setLevel(any(), any())).thenCallRealMethod();
+        List<String> modified = (List<String>) loggingResource.setLevel("a.b.c.p", Collections.singletonMap("level", "DEBUG")).getEntity();
+        assertEquals(4, modified.size());
+        assertEquals(Arrays.asList("a.b.c.p", "a.b.c.p.X", "a.b.c.p.Y", "a.b.c.p.Z"), modified);
```
How about checking whether the logger levels were actually set?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,200 @@
+@Consumes(MediaType.APPLICATION_JSON)
+public class LoggingResource {
+
+    private static final String ROOT_LOGGER_NAME = "@root";
```
Should this be public and have JavaDoc?

rhauch reviewed (commented):
Thanks, @wicknicks! A few more minor comments / suggestions below.

wicknicks committed:
KAFKA-7772: Address reviewer comments

Use standard name for Root logger
Clean up tests
Make checkstyle happy.

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
addressed your comments, @rhauch.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResource.java:
```diff
@@ -0,0 +1,207 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime.rest.admin;
```
Did you consider putting this class in the existing `org.apache.kafka.connect.runtime.rest.resources` package with the other REST resources? I know we haven't had to qualify the different type of resources, but reusing that existing package would allow us to avoid creating a package with a single class, and it would keep all REST resource classes (no matter the category or type) in the same place. WDYT?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/admin/LoggingResourceTest.java:
```diff
@@ -0,0 +1,189 @@
+        when(loggingResource.setLevel(any(), any())).thenCallRealMethod();
+        List<String> modified = (List<String>) loggingResource.setLevel("@root", Collections.singletonMap("level", "DEBUG")).getEntity();
+        assertEquals(5, modified.size());
+        assertEquals(Arrays.asList("a.b.c.p.X", "a.b.c.p.Y", "a.b.c.p.Z", "a.b.c.s.W", "root"), modified);
```
Can we also check the actual logger's levels here, just like in `setLevelTest()`?

rhauch reviewed (commented):
Thanks, @wicknicks. Just a few more very minor suggestions.

wicknicks committed:
KAFKA-7772: Move LoggingResource into resources package

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch reviewed (approved):
LGTM. Thanks, @wicknicks!

Waiting for a green build, though it does build locally for me.
